### PR TITLE
feat: change package.json config name to 'hlx'

### PR DESCRIPTION
**Pull request by alexkli**
## Description

This changes the config name `wsk` to `hlx` inside `package.json`. `wsk` is still supported for backwards compatibility.

**Please vote for the new name in comments!**

I chose `hlx` for now, but any name that makes it more generic or named after the tool sound good to me. Other ideas:
* `hlx`
* `helixDeploy`
* `serverless`
* `serverlessFunction`
* `faas`

## Motivation and Context

Helix-deploy is not just limited to openwhisk, so seeing `wsk` in the configuration when eg. using helix-deploy exclusively with AWS Lambda is quite confusing.

## How Has This Been Tested?

Unit test fixtures have been changed to `hlx`. Two new fixtures & tests have been added to test the old name `wsk` (`*-conf-wsk`).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

File: src/cli.js [hunk 1]
```diff
@@ -65,10 +73,31 @@ async function loadPlugin(name) {
...
 }
 
+function findPkgJsonConfig() {
+  // fallback to original config name 'wsk' in case anything goes wrong
+  // for safe backwards compatibility
+  let cfg = 'wsk';
+
+  try {
+    // find package.json in parent directories (like yargs does)
+    const pkg = escalade(process.cwd(), (_, names) => (names.includes('package.json') ? 'package.json' : undefined));
```
Comment (alexkli) [inline src/cli.js:83]: FYI I followed the yargs implementation: https://github.com/yargs/yargs/blob/0c95f9c79e1810cf9c8964fbf7d139009412f7e7/lib/yargs-factory.ts#L1611

File: test/build.test.js [hunk 2]
```diff
@@ -83,7 +83,9 @@ describe('Build Test', () => {
 
     await builder.run();
 
-    await assertZipEntries(path.resolve(testRoot, 'dist', 'simple-package', 'simple-name@1.45.0.zip'), [
+    const zipPath = path.resolve(testRoot, 'dist', 'simple-package', 'simple-name@1.45.0.zip');
+    assert.ok(fse.existsSync(zipPath), `Build did not produce zip file: ${zipPath}`);
```
Comment (alexkli) [inline test/build.test.js:87]: FYI This is unrelated to the PR, but when unit tests failed here, it didn't show the stack trace but only the file not found error. This extra assertion makes that nicer.

File: src/cli.js [hunk 3]
```diff
@@ -28,6 +30,12 @@ const PLUGINS = [
   WebpackBundler,
 ];
 
+const PKG_CONF = [
+  'hlx',
+  // original name for backwards compatibility
+  'wsk',
+];
```
Comment (alexkli) [inline src/cli.js:37]: Here we could add more name choices if we want to 😃

File: package.json [hunk 4]
```diff
@@ -51,6 +51,7 @@
     "archiver": "7.0.1",
     "chalk-template": "1.1.0",
     "dotenv": "16.4.5",
+    "escalade": "^3.2.0",
```
Comment (alexkli) [inline package.json:54]: This dependency is already present with `yargs`.

File: src/cli.js [hunk 5]
```diff
@@ -65,10 +73,31 @@ async function loadPlugin(name) {
...
+  }
+
+  return cfg;
+}
+
 export default class CLI {
   buildArgs(plugins) {
     this._yargs = yargs()
-      .pkgConf('wsk')
+      .pkgConf(findPkgJsonConfig())
```
Comment (alexkli) [inline src/cli.js:100]: Unfortunately yargs does not support multiple config names here. Hence I have to find the package.json first (in the same manner as yargs, walking up the tree from current dir) and see what of our candidate names is present in the file. And then tell yargs to look for that one. If none is present or things fail, we just default to `wsk` so it all behaves as before.

**Comment by justinedelson** [comment]
@alexkli would it make sense to log a warning when `wsk` is used? If it's deprecated that suggests that someday it'll be removed (although this isn't literally what 'deprecated' means it's commonly understood to mean that).

**Review by justinedelson — APPROVED** [review]

File: README.md [hunk 6]
```diff
@@ -247,20 +247,22 @@ in the `wsk` object. eg:
 }
 ```
 
+The config name changed to `hlx` in version `12.3`. The previous name `wsk` is still supported, but deprecated.
```
Comment (justinedelson) [inline README.md:250]: Maybe also add something here that says that you must use ONE or the OTHER, i.e.

```
{
    "wsk" : {
        "name": "foo"
    },
    "hlx" : {
        "somethingelse": "bar"
    },
}
```
is not supported (`wsk` will be ignored IIUC in that case).

And maybe that should result in a fatal error.

(this and my other comment are potential improvements but not impediments to merging IMO... the behavior as written is sensible and understandable)

**Comment by alexkli** [comment]
> If it's deprecated that suggests that someday it'll be removed
I have no strong opinion on a deprecation process here. Works for me to keep supporting it forever. I leave that decision to the original maintainers.

I just added the word "deprecated" in the README to put more emphasis on the new name I guess.

File: README.md (hunk 6 above)
Comment (alexkli) [reply README.md]: I added a note in the readme that `hlx` will be used if both are present.

File: package.json (hunk 4 above)
Comment (tripodsan) [inline package.json:54]:
```suggestion
    "escalade": "3.2.0",
```

File: src/cli.js [hunk 7]
```diff
@@ -65,10 +73,31 @@ async function loadPlugin(name) {
...
+  // fallback to original config name 'wsk' in case anything goes wrong
+  // for safe backwards compatibility
+  let cfg = 'wsk';
+
+  try {
+    // find package.json in parent directories (like yargs does)
+    const pkg = escalade(process.cwd(), (_, names) => (names.includes('package.json') ? 'package.json' : undefined));
+
+    if (pkg) {
+      const pkgJson = JSON.parse(readFileSync(pkg, 'utf8'));
```
Comment (tripodsan) [inline src/cli.js:86]: can we make this async?

File: test/build.test.js (hunk 2 above)
Comment (tripodsan) [inline test/build.test.js:87]:
```suggestion
    assert.ok(await fse.exists(zipPath), `Build did not produce zip file: ${zipPath}`);
```

**Review by tripodsan — APPROVED** [review]
LGTM... I don't like the synchronous I/O access, but other than that.

File: src/cli.js (hunk 7 above)
Comment (alexkli) [reply src/cli.js:86]: Yeah but then you have to make `CLI.buildArgs()` also async and maybe more downstream... and I wanted to keep the change limited. IMO it's not bad for a cli which only has "one thread" here anyway?

Comment (alexkli) [reply src/cli.js:86]: Not arguing either way, but `yargs` itself is very sync-y internally: https://github.com/yargs/yargs/blob/0c95f9c79e1810cf9c8964fbf7d139009412f7e7/lib/yargs-factory.ts#L1637

**Review by solaris007 — APPROVED** [review]

**Review by trieloff — APPROVED** [review]

**Pull request merged**